### ISSUE TITLE
How to use Vulkan related functions

### ISSUE THREAD

**Issue by Timxu2016** [ValveSoftware/openvr#346]
virtual uint32_t GetVulkanInstanceExtensionsRequired( VR_OUT_STRING() char *pchValue, uint32_t unBufferSize ) = 0;
virtual uint32_t GetVulkanDeviceExtensionsRequired( VkPhysicalDevice_T *pPhysicalDevice, VR_OUT_STRING() char *pchValue, uint32_t unBufferSize ) = 0;

I tried to print pchValue but always get nothing.can anyone help explain the right way to use these two? Thanks!

**Comment by Plagman** [comment]
I think you're supposed to call it with NULL, 0 initially and allocate a character buffer with the size of the return value, then call it again with a pointer to that character buffer and its length, and it will fill it for you.

Then you feed the resulting extensions directly into your device/instance creation functions. Note that it might tell you that it doesn't need to enable any extensions, depending on platform/driver/GPU.

See this page for more information:

https://github.com/ValveSoftware/openvr/wiki/Vulkan

**Comment by Timxu2016** [comment]
Thank your for your help!! I tried to calll like this:  
       uint32_t buffersize;
       buffersize = vr::VRCompositor()->GetVulkanInstanceExtensionsRequired(nullptr, 0);
       //vr::VRCompositor()->GetVulkanInstanceExtensionsRequired(nullptr, buffersize );
	char* pValue = (char*)malloc(buffersize);
	vr::VRCompositor()->GetVulkanInstanceExtensionsRequired(pValue, buffersize);
But still it doesn't work. buffersize always return 0. :(

**Comment by zao** [comment]
The comments in the source code seem a bit cut off, but isn't it possible that the extensions needed are the empty set, and thus you get a string size of zero?

**Comment by Plagman** [comment]
What platform/driver/GPU are you on?

**Comment by Timxu2016** [comment]
@zao maybe, but i'm still curious if it can return me some results to know what it looks like.
@Plagman GTX980 
![default](https://cloud.githubusercontent.com/assets/19217893/21561249/3398ae10-cea5-11e6-94b4-d46559662808.png)

**Comment by Plagman** [comment]
Hi, it looks like you have an old driver. Can you get the latest driver from NVIDIA and try again? Looks like you have 364.72 but it should be up to the 376 series now.

**Comment by Timxu2016** [comment]
@Plagman Thank you very much, that's my missing point. 
Now i can get VK_NV_external_memory_capabilities from GetVulkanInstanceExtensionsRequired(),
but still 0 from GetVulkanDeviceExtensionsRequired. Maybe i don't need to enable any extra device extensions?

`	uint32_t buffersize2;
	buffersize2 = vr::VRCompositor()->GetVulkanDeviceExtensionsRequired(info.gpus[0], NULL, 0);
	char* pValue2 = (char*)malloc(buffersize2);
	vr::VRCompositor()->GetVulkanDeviceExtensionsRequired(info.gpus[0], pValue2, buffersize2);`

Anyway, I tried next to build VRVulkanTextureData_t for submit(). The image layout and format is set elsewhere as the wiki required.

`	vr::VRVulkanTextureData_t leftEye = {};
	//leftEye.m_nImage = (uint64_t)info.current_buffer;
	//leftEye.m_nImage = (uint64_t)info.buffers[info.current_buffer].view;
	leftEye.m_nImage = (uint64_t)info.buffers[info.current_buffer].image;
	leftEye.m_pDevice = info.device;
	leftEye.m_pInstance =info.inst;
	leftEye.m_pQueue = info.graphics_queue;
	leftEye.m_nQueueFamilyIndex = info.graphics_queue_family_index;
	leftEye.m_nWidth = info.width;
	leftEye.m_nHeight = info.height;
	leftEye.m_nFormat = info.format;
	leftEye.m_nSampleCount = 1;

	vr::Texture_t leftEyeTexture = { (void*)&leftEye, vr::TextureType_Vulkan, vr::ColorSpace_Gamma };
	vr::VRCompositor()->Submit(vr::Eye_Left, &leftEyeTexture);`

I tried many formats of m_nImage, but neither worked.  Am I missing anything again?  I would really appreciate if you can help me again : )

**Comment by Plagman** [comment]
I think you're right that there is not a device extension needed right now, although it might change in the future, so you should still wire it up. Are you submitting both left and right eyes and still not getting any result?

**Comment by Timxu2016** [comment]
@Plagman Yes,  I just paste lefteye code here, but did submit to both eyes with the same image.

**Comment by Plagman** [comment]
Would it be easy for you to send us a test app that reproduces the problem? That would probably be the fastest way to troubleshoot on this end.
